src/MPMCQueue.NET/MPMCQueue.cs

```diff
@@ -1,91 +1,117 @@
 ﻿using System;
 using System.Runtime.InteropServices;
 using System.Threading;
+using System.Runtime.CompilerServices;
 
 namespace MPMCQueue.NET
 {
-    [StructLayout(LayoutKind.Explicit, Size = 192, CharSet = CharSet.Ansi)]
+    [StructLayout(LayoutKind.Explicit, Size = 336)]
     public class MPMCQueue
     {
-        [FieldOffset(0)]
-        private readonly Cell[] _buffer;
-        [FieldOffset(8)]
-        private readonly int _bufferMask;
-        [FieldOffset(64)]
-        private int _enqueuePos;
-        [FieldOffset(128)]
-        private int _dequeuePos;
+        /// <summary>
+        /// 128 bytes cache line already exists in some CPUs.
+        /// </summary>
+        /// <remarks>
+        /// Also "the spatial prefetcher strives to keep pairs of cache lines in the L2 cache."
+        /// https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding
+        /// </remarks>
+        internal const int SAFE_CACHE_LINE = 128;
 
+        [FieldOffset(SAFE_CACHE_LINE)]
+        private readonly Cell[] _enqueueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE + 8)]
+        private volatile int _enqueuePos;
+
+        // Separate access to buffers from enqueue and dequeue.
+        // This removes false sharing and accessing a buffer 
+        // reference also prefetches the following Pos with [(64 - (8 + 4 + 4)) = 52]/64 probability.
+
+        [FieldOffset(SAFE_CACHE_LINE * 2)]
+        private readonly Cell[] _dequeueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE * 2 + 8)]
+        private volatile int _dequeuePos;
```

Comment:
Hello :)
I wonder if the observed improved performance really comes from having two copies of the same read-only variable, or rather from the fact that the frequently read fields are no longer in the same cache line as the fields being modified frequently.
I mean: I can see how constantly writing to a cache line can trash it and cause false sharing issues, but I fail to see how having multiple readers is a problem - the cache line should be in (S)hared state in MESI protocol, right?
I see that in the old code, the read-only `_bufferMask` had offset 8, and write-intensive `_enqueuePos` had offset 64, but does it really mean they are guaranteed to be in different cache lines?
What if the `MPMCQueue` instance itself is not aligned at cache line boundary?
Could it then be that `__bufferMask` is at the start of the cache line and `__enqueuePos` at its end?
And if the Facebook's observation that we should rather care about 128 rather than 64 units is correct, then the same applies to `__buffer` and `__enqueuePos` (and other pairs of "read-only" and "write-havy" fields as many of them are closer than 128 bytes to each other).
In description of PR you give one more probable reason of why duplication helps: data needed to enqueue is in one line, and data needed for dequeue is in one (other) line, which creates nice "packages" of self-sufficient data. I can see how it can help: it requires one fetch instead of two, in case data was missing from cache. But how frequent is the situation of cache miss in a tight loop of benchmark?
I would appreciate someone educate me on the most probable explanation (ideally: backed with experiment) why this really helps :)

Reply:
> Could it then be that __bufferMask is at the start of the cache line and __enqueuePos at its end?

With probability 8/64 yes. .NET aligns objects to 8 bytes, not cache line. And yes, if anything is modified on a cache line all other threads will need to reload that line, so modifying enqueuePos could interfere with Dequeue reading bufferMask with 8/64 probability.